pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,469 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitortool
+
+import (
+	"math/rand"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+	corev1 "k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha1"
+)
+
+var (
+	icmpSeq uint32
+	// #nosec G404: random number generator not used for security purposes.
+	icmpEchoID = rand.Int31n(1 << 16)
+)
+
+const (
+	ipv4ProtocolICMPRaw = "ip4:icmp"
+	ipv6ProtocolICMPRaw = "ip6:ipv6-icmp"
+	protocolICMP        = 1
+	protocolICMPv6      = 58
+)
+
+// getICMPSeq returns the next sequence number as uint16,
+// wrapping around to 0 after reaching the maximum value of uint16.
+func getICMPSeq() uint16 {
+	// Increment the sequence number atomically and get the new value.
+	// We use atomic.AddUint32 and pass 1 as the increment.
+	// The returned value is the new value post-increment.
+	newVal := atomic.AddUint32(&icmpSeq, 1)
+
+	return uint16(newVal)
+}
+
+// NodeLatencyMonitor is a tool to monitor the latency of the Node.
+type NodeLatencyMonitor struct {
+	// latencyStore is the cache to store the latency of each Nodes.
+	latencyStore *LatencyStore
+	// latencyConfigChanged is the channel to notify the latency config changed.
+	latencyConfigChanged chan LatencyConfig
+	// isIPv4Enabled is the flag to indicate whether the IPv4 is enabled.
+	isIPv4Enabled bool
+	// isIPv6Enabled is the flag to indicate whether the IPv6 is enabled.
+	isIPv6Enabled bool
+
+	// nodeName is the name of the current Node, used to filter out the current Node from the latency monitor.
+	nodeName                   string
+	nodeInformer               coreinformers.NodeInformer
+	nodeLatencyMonitorInformer crdinformers.NodeLatencyMonitorInformer
+}
+
+// LatencyConfig is the config for the latency monitor.
+type LatencyConfig struct {
```

Comment:
better to make it private

Reply:
Ok, this struct is not used externally, private is a better choice.